### ISSUE TITLE
Support RavenDB with Cluster Wide Transactions

### ISSUE THREAD

**Issue by kblooie** [Particular/NServiceBus.RavenDB#436]
We're currently upgrading to Raven 4.2 from 3 and we really wanted to try to have higher availability in our sagas. @DavidBoike mentioned in #412 that issues were encountered with Cluster Wide Transactions.  Could you provide details on what issues were encountered? 

I have a block of hours I am dedicating to getting an implementation running. Any details would be appreciated. If you'd like a PR, please let me know.

**Comment by DavidBoike** [comment]
@kblooie 	would you have some time to chat about this? I'll send you an email.

**Comment by rvaridel** [comment]
@DavidBoike, We're looking to use the Transactional Outbox feature but have a RavenDB 4.2 multi-master cluster topology which we'd like to retain for automatic failover etc. I noticed the PR to add support for cluster-wide transactions is marked as WIP/Blocked. 

Are you perhaps able to provide details of the outstanding issues and if you need any help to get this across the line?

**Comment by kblooie** [comment]
We implemented an outbox & saga persistence implementation using
cluster-wide transactions, but decided to go a different way. The biggest
problem we faced was that cluster-wide transactions replaces optimistic
concurrency in RavenDB AND the transactions are bound by the cluster-wide
transaction keys you provide. If you modify a non-nservicebus document
during your NSB message handling and also save the same document around the
same time in a Web endpoint using traditional optimistic concurrency, your
web changes will most likely get lost. This is because Raven by design
gives precedence to cluster-wide transactions over its optimistic
concurrency implementation. You could get around this by using cluster-wide
transactions for all your business data, but that would be complex, slow,
and defeats the point of having a cluster with fail-over.

In the end, we chose instead to relax the topology check that NSB does and
run using optimistic concurrency with fail-over in our multi-node cluster.
We found that we much preferred to handle the consistency issues of double
handling messages and saga conflicts in rare fail-over scenarios than
dealing with cluster-wide transactions across our system.

@David Boike <david.boike@particular.net> We're currently running our own
persistence with small changes that allow us to disable to topology check.
We've included opting into this using config extension methods. Would you
accept a pull request to allow people to run in this manner if they choose
to do so?

Jonathan

**Comment by DavidBoike** [comment]
Hi @rvaridel,

At the time we were just trying to upgrade our support to include RavenDB 4. We ran into issues, and given that we hadn't supported multi-master clusters on RavenDB 3.x, we implemented for RavenDB 4 as you see it now - with support for a single cluster leader and multiple followers.

There are some technical challenges around the use of compare-exchange, were we to do it. For example, in order for the Outbox to work, a compare exchange record would need to be created, but unlike documents, you can't delete compare exchange records with a KeyStartingWith expression. Without a way to clean them up they'd eventually fill your disk. RavenDB has indicated there is a new document refresh feature that "touches" a document after a metadata-driven period of time, which allows you to then subscribe to that change and delete the compare exchange at that time, but that's only been added in very recent versions of the server (4.2.2 and up, I _think_, and not exposed in Studio until 4.2.4).

But @kblooie touched on the bigger problem. Opting in to a cluster-wide transaction in the message handling pipeline is a Really Big Deal® with ramifications for every bit of code executed in message handler. It's difficult to guide users to the pit of success here as bugs introduced by not having proper compare-exchange usage in common data access are going to end up being very weird and hard to track down.

So I don't know if it's a matter of implementation or even prioritization. I'm not sure that we're comfortable even attempting to support cluster-wide transactions right now.

@kblooie I'm curious - are you then, for example, pre-emptively creating conflict resolver code for every saga type?

**Comment by rvaridel** [comment]
Hi @DavidBoike,

Considering @kblooie issues with CWTs and design drawbacks, I think we'll embrace the eventual consistency model and also simply enable Optimistic Concurrency checking to prevent overwrites. Our implementation will only modify a document from within an NSB handler so I think our risk of a conflict scenario is quite low here. 

I also support @kblooie's suggestion of having an extension method or a mechanism to bypass the topology check or perhaps enforce that Optimistic Concurrency is enabled if using a multi-master cluster topology. 

What are your thoughts?

**Comment by DavidBoike** [comment]
I'm actually very curious about how @kblooie is doing that. :wink:

If it's possible to do it as a workaround, I'd rather document that process here as a "Use At Your Own Risk" rather than adding an API option that, if not used _extremely_ carefully, could lead to a message loss or data corruption scenario.

**Comment by kblooie** [comment]
We're running a forked version of this repo, with a few customizations that I will send a PR for soon.  I've just submitted a small PR that allows users to disable the topology check on startup #446.

@DavidBoike We've written conflict resolvers for a few key sagas that have high contention to prove that we could, but have mostly decided to manually resolve conflicts to see where we have issues. We've only had a handful of conflicts in the months that we've released and the conflicts fail as expected. This is on a system with many sagas and over 100 endpoints and thousands of messages per day. 

For the outbox. We also manually resolve conflicts using the management UI to allow us to assess the impact to our system. In all our cases, they could be safely ignored. Writing a conflict resolver to ignore all instances that can be ignored would be simple.

Since the failover logic in Raven 4 now limits jitter when a node comes back up, instances of conflicts are dramatically reduced. Though I think that running in a single node will allow users to default into the pit of success, it's not practical in highly available systems. I think that users should be allowed to choose how consistent they want to be using the override in the provided PR. You could perhaps add warnings or change the name of the extension method to better communicate consistency issues that they're signing up for.

**Comment by DavidBoike** [comment]
We're going to take a wait-and-see position here. As I said in https://github.com/Particular/NServiceBus.RavenDB/pull/446#issuecomment-548877660:

> We need to be extremely cautious about taking down the safety rails and opening users up to document conflicts.

I'm particularly concerned that there may be systems that, due to higher saga contention or whatever, would be more likely to produce document conflicts that, given they don't own the document schema (outbox, subscriptions, etc.) or only partially own them (sagas) it would be difficult to deal with those conflicts in practice.

**Comment by DavidBoike** [comment]
FYI, we have new information from Oren that only a single-node "cluster" (i.e. not a cluster) is a totally safe configuration.

Me:

> Are you saying that in a cluster with 1 Member/Leader and multiple Watchers, it would be possible for 2 concurrently processing messages to read data from different database instances, successfully update that data differently, and result in a conflict if cluster-wide transactions and CompareExchange operations are not used?

Oren:

> Yes, that is correct.

We have opened an issue internally to investigate and determine what to do.

**Comment by kbaley** [comment]
We're investigating this for the current RavenDB enhancement release. Still too early to get hopes up but we might be able to pull this off if we can detect whether cluster-wide transactions are used.

**Comment by mauroservienti** [comment]
A first-round of an investigation proves that it's possible to implement cluster-wide transactions for saga data. At a first look, it also seems that there is no need to implement cluster-wide support for the outbox persister; the reason being that if the saga persistence fails due to the transaction being rejected the outbox is never invoked. If the saga persister succeeds in acquiring a cluster-wide compare exchange then all subsequent outbox operations are append-only stuff that no other endpoint instances will be doing anyway.

Things to further investigate:

- the persister allows the user to provide a custom session, this means that we have no control over the session creation and the requirement in the pipeline is to understand if the provided session is set to use cluster-wide transactions in which case the saga persister has to use compare exchange documents; otherwise, it's just regular RavenDB business;
- does the outbox persister need to use cluster-wide transactions when setting messages as dispatched?
- Is there an opportunity to use cluster-wide transactions as a lock kind of mechanism when beginning the outbox transaction?
- What's the impact, if any, of cluster-wide transactions on the saga pessimistic locking opt-in support?
- How do we build acceptance tests to make sure that the behavior is correct?
- As [pointed out](https://github.com/Particular/NServiceBus.RavenDB/issues/436#issuecomment-547182078) we need to carefully document the impact of using cluster-wide transactions

> The biggest problem we faced was that cluster-wide transactions replace optimistic
concurrency in RavenDB AND the transactions are bound by the cluster-wide
transaction keys you provide. If you modify a non-NServiceBus document
during your NSB message handling and also save the same document around the
same time in a Web endpoint using traditional optimistic concurrency, your
web changes will most likely get lost. This is because Raven by design
gives precedence to cluster-wide transactions over its optimistic
concurrency implementation. You could get around this by using cluster-wide
transactions for all your business data, but that would be complex, slow,
and defeats the point of having a cluster with fail-over.

This probably means that if users are setting up the endpoint to use cluster-wide transactions for saga and while handling messages they also touch business data, they are exposed and forced to use cluster-wide transactions API for their business data as well, at the very minimum in the context of the saga.

**Comment by mauroservienti** [comment]
We raised https://github.com/Particular/NServiceBus.RavenDB/pull/624. So far we have:

- cluster wide transactions support for sagas
- cluster wide transactions support for subscriptions
- next will be support for the outbox storage

We made a conscious decision, for now, to enable cluster-wide transactions support globally through persistence configuration. like the following:

```
var persistenceConfig = endpointConfig.UsePersistence<RavenDBPersistence>();
persistenceConfig.UseClusterWideTransactions();
```

The above means that all operations performed when using the NServiceBus session will go through cluster wide transactions in a cluster. We're evaluating if there are options and good reasons to relax this constraint and allow transactions to be configured on a session by session basis.
